src/rulesets/Base/indexing.jl

```diff
@@ -125,6 +207,21 @@ function frule((_, ẋ, v̇), ::typeof(setindex!), x::AbstractArray, v, inds...)
     return setindex!(x, v, inds...), setindex!(ẋ, v̇, inds...)
 end
 
+#####
+##### unsafe_getindex
+#####
+
+# This is called by e.g. `iterate(1:0.1:2)`,
+# and fixes https://github.com/FluxML/Zygote.jl/issues/1247
+# Only needs to accept AbstractRange, but AbstractVector makes testing easier.
+
+function frule((_, ẋ), ::typeof(Base.unsafe_getindex), x::AbstractVector, i::Integer)
+    return Base.unsafe_getindex(x, i), getindex(ẋ, i)
+end
+
+function rrule(cfg::RuleConfig{>:HasReverseMode}, ::typeof(Base.unsafe_getindex), x::AbstractVector, i::Integer)
```

Comment:
This doesn't need any mode does it?
```suggestion
function rrule(cfg::RuleConfig, ::typeof(Base.unsafe_getindex), x::AbstractVector, i::Integer)
```
Not sure why we aren't just calling the `rrule` for get infact?
Or putting a `Union{typeof(getindex), typeof(Base.getindex))` in the function arg on them.
(might even be able to stick `view` in that union too?)

Reply:
I guess I thought the first Zygote use of this might leave its own rule intact for scalar indexing, to make OneElement for that, and use CR for all others. In which case `rrule_via_ad` here will call that.

More generally it may want to do other more efficient things for indexing ranges. Or we may want to do that here, and remove this entirely.